apps/gallery/js/nfc.js

```diff
@@ -10,26 +10,62 @@
  * When the UI changes so that no sharable object is visible, call unshare().
  */
 var NFC = {
-  share: function share(f) {
-    // Enable NFC sharing by defining a "peerready" callback that
-    // calls sendFile() to transfer the current image or video to the peer
+  shrinkingUI: null,
+  share: function share(f, options) {
     if (navigator.mozNfc) {
-      navigator.mozNfc.onpeerready = function(event) {
-        if (typeof f === 'function') {
-          var promise = f();
-          promise.then(function(file) { event.peer.sendFile(file); });
-        }
-        else {
-          event.peer.sendFile(f);
-        }
-      };
+      // Bring shrinkingUI to gallery's activity window.
+      // XXX: It's a workaround for bug 1163957. The ideal fix should be done in
+      //      bug 1109495, but the changes should be applied to all NFC
+      //      supported apps.
+      if (window.ShrinkingUI && options) {
+        var peer = null;
+        NFC.shrinkingUI = new window.ShrinkingUI(options.foregroundElement,
+          options.backgroundElement);
+        navigator.mozNfc.onpeerfound = function(event) {
+          peer = event.peer;
+          NFC.shrinkingUI.start();
+          // It will tell gecko that gallery app will handle nfc onpeer event,
+          // or onpeerlost event would be triggered immediately.
+          event.preventDefault();
+        };
+        navigator.mozNfc.onpeerlost = function() {
+          peer = null;
+          NFC.shrinkingUI.stop();
+        };
+        window.addEventListener('shrinking-sent', function shrinkingSent() {
+          if (!peer) {
+            return;
+          }
+          NFC.sendFile(f, peer);
+        });
+      } else {
+        // Enable NFC sharing by defining a "peerready" callback that
+        // calls sendFile() to transfer the current image or video to the peer
+        navigator.mozNfc.onpeerready = function(event) {
+          NFC.sendFile(f, event.peer);
+        };
+      }
     }
   },
 
+  sendFile: function(f, peer) {
+    if (typeof f === 'function') {
+      var promise = f();
+      promise.then(function(file) { peer.sendFile(file); });
+    }
+    else {
+      peer.sendFile(f);
+    }
+    NFC.shrinkingUI && NFC.shrinkingUI.stop();
```

Comment:
Is this okay even in the promise case above? Or should you wait to call stop until the promise resolves?